### PR TITLE
Adds an example markdown design doc template

### PR DESCRIPTION
**Pull request by pwolfram**
This merge adds a template design document for MPAS-Analysis contributions.

### CODE REVIEW

**Comment by pwolfram** [comment]
@xylar I'm happy deleting "Algorithms" as @milenaveneziani proposes but think it may be useful in some cases and we could also make it "Algorithms (optional)". Do you have a preference? If we delete it, someone could always add it back if need and that person is probably going to be one of us. Hence, I'm leaning towards its removal.

**Comment by xylar** [comment]
@milenaveneziani wrote:
> I am afraid that requiring design documents for any analysis code will greatly reduce the already low level of participation.
At least, I'd say let us make it simpler: for example, the algorithm will probably be easy in most cases. Could we just have Design and Implementation, in which one could include a brief description of the algorithm at the top?

I think the template is good, but I'm fine with putting the Algorithms section as optional.  I realize it's a pain to do a design document.  That's the chance for the rest of the team to give feedback, though, before things get too far.  Part of making this process not too daunting and time consuming is being careful in our *response* to design documents to make it feel like a useful process as opposed to just steering someone away from their own ideas and plans and toward our own or just an endless process with many iterations and far too much discussion.  Both of these have felt true to me sometimes when I make design documents, but I have also got extremely valuable feedback that has kept me from wasting a lot of time.

File: design_docs/template.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<h1> Title: Some descriptive title <br>
+ACME Ocean and Ice Team <br>
+date: x - xx - xxxx <br>
+</h1>
+<h2> Summary </h2>
+The purpose of this section is to summarize what capability is to be added to the ACME Ocean and Ice system through this design process. It should be clear what new code will do that the current code does not. Summarizing the primary challenges with respect to software design and implementation is also appropriate for this section. Finally, this statement should contain general statement with regard to what is “success.”
```
Comment (xylar) [inline design_docs/template.md]: Might as well change "ACME Ocean and Ice" to "MPAS-Analysis"

File: design_docs/template.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+<h1> Requirements </h1>
+
+<h2> Requirement: name-of-requirement-here <br>
+Date last modified: 2011/01/05 <br>
+Contributors: (add your name to this list if it does not appear)
+</h2>
+
+
+Each requirement is to be listed under a ”section” heading, as there will be a one-to-one correspondence between requirements, design, proposed imple- mentation and testing. Requirements should not discuss technical software issues, but rather focus on model capability. To the extent possible, require- ments should be relatively independent of each other, thus allowing a clean design solution, implementation and testing plan.
```
Comment (xylar) [inline design_docs/template.md]: Fix the hyphenation in "imple- mentation" and "require- ments"

File: design_docs/template.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+<h1> Algorithmic Formulations </h1>
```
Comment (xylar) [inline design_docs/template.md]: Add "(optional)"

File: design_docs/template.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+
+<h1> Design and Implementation </h1>
+<h2> Implementation: short-desciption-of-implementation-here <br>
+Date last modified: 2011/01/05 <br>
+Contributors: (add your name to this list if it does not appear)
+</h2>
+
+
+This section should detail the plan for implementing the design solution for requirement XXX. In general, this section is software-centric with a focus on software implementation. Pseudo code is appropriate in this section. Links to actual source code are appropriate. Project management items, such as svn branches, timelines and staffing are also appropriate. How do we typeset pseudo code?
```
Comment (xylar) [inline design_docs/template.md]: Change "svn" to "git"

File: design_docs/template.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+
+<h1> Testing </h1>
+<h2> Testing and Validation: short-desciption-of-testing-here <br>
+Date last modified: 2011/01/05 <br>
+Contributors: (add your name to this list if it does not appear)
+</h2>
+
+
+How will XXX be tested? i.e. how will be we know when we have met requirement XXX. Will these unit tests be included in the ongoing going forward?
```
Comment (xylar) [inline design_docs/template.md]: Maybe add something here specifically about CI and the testing infrastructure.

File: design_docs/template.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+<h1> Title: Some descriptive title <br>
+ACME Ocean and Ice Team <br>
```
Comment (xylar) [inline design_docs/template.md]: Maybe should be the MPAS-Analysis team or just the author(s) name(s)?

**Comment by xylar** [comment]
I created a wiki page to see what the options might be.  
https://github.com/MPAS-Dev/MPAS-Analysis/wiki/Design-Document:-Generalized-Horizontal-Interpolation
This didn't seem to work very well in the sense that there aren't any facilities for discussing the design document within the wiki.

An alternative approach would be to open an issue with the title "Design Document: XXX" and description that is the design document.  I can also try that approach to see if it works better.

**Comment by xylar** [comment]
Take a look at #88 and see if you like this approach to posting and commenting on design documents.  If not, I'll close the issue.  I'll also clean up the various other copies of this design document elsewhere once we decide what we would like to do.

**Comment by pwolfram** [comment]
@xylar, the problem with using an issue for the design document is that we can't comment on the design document line-by line or keep track of changes (e.g., via separate commits) if desired.  Being able to comment line-by-line is really helpful and I would argue necessary for this process.  Also, issues lack history.  Having them in a PR is helpful because you get comments like you would for issues but you also have a product that gets permanently merged into the repo.

**Comment by pwolfram** [comment]
@xylar, I've made changes to the template and pushed them so this should be ready for another round of review.

Thoughts on this design-doc template @milenaveneziani?

File: design_docs/template.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
+<h1> Title: Some descriptive title <br>
+MPAS-Analysis Team <br>
+date: x - xx - xxxx <br>
+</h1>
+<h2> Summary </h2>
+The purpose of this section is to summarize what capability is to be added to
+the ACME MPAS-Analysis system through this design process. It should be clear
```
Comment (xylar) [inline design_docs/template.md]: I would suggest deleting ACME here since MPAS-Analysis should be able to live outside of ACME

Comment (xylar) [inline design_docs/template.md]: Sorry if I missed this from before.

**Comment by xylar** [comment]
@pwolfram, I agree that it makes sense to do design document as PRs for all the reasons you point out.  I will move my two current docs to be PRs as soon as I find time.

File: design_docs/template.md (hunk 7 above)
Comment (pwolfram) [inline design_docs/template.md]: Thanks for catching this.  Will fix and then should be ready to merge assuming everyone is happy with this template.  yes/no on this @milenaveneziani?

**Comment by pwolfram** [comment]
@xylar, I've edited the design doc as requested and unless @milenaveneziani has objections it should be ready to merge.

**Comment by milenaveneziani** [comment]
Looks good @pwolfram

**Pull request merged**